File: kata02.rb
class DateFormatter
  MONTH_NAMES = %W{nil January February March April May June July August September October November December}

  def initialize(input_date)
    @year, @month, @day = input_date.split("-").map(&:to_i)
  end

  attr_reader :month, :day, :year

  def print_us
    print sprintf "US: %02<month>d/%02<day>d/%04<year>d ", {month: month, day: day, year: year}
  end

  def print_uk
    print sprintf "UK: %02<day>d/%02<month>d/%04<year>d ", {month: month, day: day, year: year}
  end

  def print_spoken
    print "Spoken: #{day}#{ordinal} of #{month_names} #{year} "
  end

  def print_all
    print_us
    print_uk
    print_spoken
  end

  private
  def ordinal
    case true
    when [1,21,31].include?(day)
      "st"
    when [2,22].include?(day)
      "nd"
    when [3,23].include?(day)
      "rd"
    else
      "th"
    end
  end

  def month_names
    MONTH_NAMES[month]
  end
end

DateFormatter.new(gets.chomp).print_all

File: kata02.py
#!/usr/bin/env python3

class Date():
    @staticmethod
    def from_str(plain):
        from re import match
        (year, month, day) = match(r"(\d\d\d\d)-(\d\d)-(\d\d)", plain).groups()
        return Date(int(year), int(month), int(day))

    def __init__(self, year, month, day):
        self.year = year
        self.month = month
        self.day = day
        self.month_name = "January February March April May June July August September October November December".split(" ")[self.month - 1]
        self.day_suffix = {1: "st", 2: "nd", 3: "rd"}.get(self.day % 10, "th")

    def __format__(self, format_spec):
        return format_spec.format(y=self.year, m=self.month, d=self.day, M=self.month_name, s=self.day_suffix)

    def to_us(self):
        return format(self, "{m:2}/{d:2}/{y:4}")

    def to_gb(self):
        return format(self, "{d:2}/{m:2}/{y:4}")

    def to_spoken(self):
        return format(self, "{d}{s} of {M} {y}")

if __name__ == "__main__":
    date = Date.from_str(input("Enter ISO Date: "))
    print("GB: ", date.to_gb())
    print("US: ", date.to_us())
    print("Spoken: ", date.to_spoken())
